drivers: clock_control: nrf54h: obtain oscillator parameters from BICR

gmarull opened the pull request:
BICR should be taken as the source of truth for certain oscillator-related parameters, such as:

- LFOSC
  - Accuracy
  - Mode
  - Load capacitance
  - Startup time
- HFXO
  - Mode
  - Load capacitance
  - Startup time

Currently, these were replicated in Devicetree, making it possible to have a non-fully functional system due to mismatches.

gmarull committed:
drivers: clock_control: nrf54h-hfxo: use values from BICR

The real, applicable and trusted values are the ones flashed into BICR.
So, drop DT properties that replicate BICR and use runtime reads to BICR
instead.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
drivers: clock_control: nrf54h-lfclk: use values from BICR

The real, applicable and trusted values are the ones flashed into BICR.
So, drop DT properties that replicate BICR and use runtime reads to BICR
instead.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

anangl commented on drivers/clock_control/clock_control_nrf2_hfxo.c:
```diff
@@ -166,6 +166,8 @@ static int init_hfxo(const struct device *dev)
 		return rc;
 	}
 
+	dev_data->start_up_time = K_USEC(NRF_APPLICATION_BICR_NS->HFXO.STARTUPTIME);
```
`NRF_APPLICATION_BICR_NS` is defined only for the Application core, so this will not compile for other cores. I guess we'll need to use some hack instead of this symbol until MDK is updated.
Also, I think an error should be returned here if this value turns out to be unconfigured.

gmarull replied:
My idea was to define BICR in DT, so we'd obtain it from there. I guess it is accessible from other cores, like radio @hubertmis ?

Good point about handling uninitialized value.

hubertmis replied:
Yes, UICR (including BICR) is readable by all CPUs

anangl commented on drivers/clock_control/clock_control_nrf2_lfclk.c:
```diff
@@ -56,49 +55,52 @@ static const struct clock_options {
 	uint32_t fixed_frequency;
 };
 
+static int lfosc_get_accuracy(uint16_t *accuracy)
```
It would be better to place this function next to `lfclk_init()` which is its only user.

anangl commented on drivers/clock_control/clock_control_nrf2_lfclk.c:
```diff
@@ -234,6 +236,45 @@ static int lfclk_init(const struct device *dev)
+		clock_options[LFCLK_NUM_OPTS - 2].accuracy = dev_data->max_accuracy;
+		clock_options[LFCLK_NUM_OPTS - 2].precision = 0;
+		clock_options[LFCLK_NUM_OPTS - 2].src = NRFS_CLOCK_SRC_LFCLK_XO_EXT_SQUARE;
+		break;
```
In this case, when `clock_options[LFCLK_NUM_OPTS - 1]` is empty, this loop: https://github.com/zephyrproject-rtos/zephyr/blob/dded562dc301c72757a123b6002f98fe74b36126/drivers/clock_control/clock_control_nrf2_lfclk.c#L164-L172 should not look into that item. I think we need `dev_data->clock_option_cnt` to use for that loop instead of `ARRAY_SIZE(clock_options)`.

gmarull replied:
good point

anangl commented on drivers/clock_control/clock_control_nrf2_lfclk.c:
```diff
@@ -224,6 +193,40 @@ static int api_get_rate_lfclk(const struct device *dev,
 
+static int lfosc_get_accuracy(uint16_t *accuracy)
+{
+	switch (NRF_APPLICATION_BICR_NS->LFOSC.ACCURACY) {
```
```suggestion
	switch (FIELD_GET(NRF_APPLICATION_BICR_NS->LFOSC.LFXOCONFIG,
			  BICR_LFOSC_LFXOCONFIG_ACCURACY_Msk)) {
```

anangl commented on drivers/clock_control/clock_control_nrf2_lfclk.c:
```diff
@@ -234,6 +237,50 @@ static int lfclk_init(const struct device *dev)
+		return ret;
+	}
+
+	switch (NRF_APPLICATION_BICR_NS->LFOSC.MODE) {
```
```suggestion
	switch (FIELD_GET(NRF_APPLICATION_BICR_NS->LFOSC.LFXOCONFIG,
			  BICR_LFOSC_LFXOCONFIG_MODE_Msk)) {
```

gmarull committed:
dts: bindings: misc: add nordic,nrf-bicr

Add binding for the Nordic nRF BICR memory.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: common: nordic: nrf54h20: define BICR node

BICR (Board Information Configuration Registers) are located within the
application UICR region (ref. MRAM mapping, table 38).

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
drivers: clock_control: nrf54h-common: add utility to obtain LFOSC acc

Add a utility function to obtain LFOSC accuracy in PPM from BICR.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
drivers: clock_control: nrf54h-fll16m: use values from BICR

The real, applicable and trusted values are the ones flashed into BICR.
So, drop DT properties that replicate BICR and use runtime reads to BICR
instead.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

hubertmis commented on drivers/clock_control/clock_control_nrf2_fll16m.c:
```diff
@@ -28,21 +28,13 @@ BUILD_ASSERT(DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) == 1,
 
 #define FLL16M_HAS_LFXO DT_NODE_HAS_STATUS_OKAY(FLL16M_LFXO_NODE)
```
I think this information should also be taken from BICR

hubertmis commented on drivers/clock_control/clock_control_nrf2_fll16m.c:
```diff
@@ -28,21 +28,13 @@ BUILD_ASSERT(DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) == 1,
 #define FLL16M_HAS_LFXO DT_NODE_HAS_STATUS_OKAY(FLL16M_LFXO_NODE)
 
-#define FLL16M_LFXO_ACCURACY DT_PROP(FLL16M_LFXO_NODE, accuracy_ppm)
 #define FLL16M_HFXO_ACCURACY DT_PROP(FLL16M_HFXO_NODE, accuracy_ppm)
```
And this one

gmarull replied:
I don't see any accuracy setting for HFXO in BICR

hubertmis replied:
Right. My mistake. So the accuracy should stay here.

gmarull commented:
DNM: will likely require reflashing BICR as some parameters were never written there.

bjarki-andreasen reviewed (dismissed):
It pains me to see properties related to specific peripheral instances not be part of the peripherals descriptions themselves, but that is actually how this platform works...

masz-nordic commented on soc/nordic/validate_base_addresses.c:
```diff
@@ -330,6 +330,7 @@ CHECK_DT_REG(uart136, NRF_UARTE136);
 CHECK_DT_REG(uart137, NRF_UARTE137);
 CHECK_DT_REG(uicr, NRF_UICR);
 CHECK_DT_REG(cpuapp_uicr, NRF_APPLICATION_UICR);
+CHECK_DT_REG(bicr, NRF_APPLICATION_BICR_NS);
```
```suggestion
CHECK_DT_REG(bicr, NRF_APPLICATION_BICR);
```

gmarull replied:
fixed

hubertmis commented on drivers/clock_control/clock_control_nrf2_fll16m.c:
```diff
@@ -51,7 +44,6 @@ static const struct clock_options {
 		.mode = FLL16M_MODE_OPEN_LOOP,
 	},
 	{
-		.accuracy = FLL16M_CLOSED_LOOP_ACCURACY,
 		.mode = FLL16M_MODE_CLOSED_LOOP,
 	},
```
The closed loop should have been removed. I think there was a task for this. You can remove it now and simplify the code.

gmarull replied:
same

gmarull commented on drivers/clock_control/clock_control_nrf2_lfclk.c:
```diff
@@ -227,13 +193,66 @@ static int api_get_rate_lfclk(const struct device *dev,
+
+		switch (lfosc_mode) {
+		case NRF_BICR_LFOSC_MODE_CRYSTAL:
+			clock_options[LFCLK_MAX_OPTS - 2].accuracy = dev_data->max_accuracy;
```
Note I preserved previous impl (if I got it right). I suggest further adjustments are done afterward.
